core/node/provider.go

```diff
@@ -5,145 +5,151 @@ import (
-		sys, err := batched.New(cr, q,
-			batched.ReproviderInterval(reprovideInterval),
-			batched.Datastore(repo.Datastore()),
-			batched.KeyProvider(keyProvider))
+func ProviderSys(reprovideInterval time.Duration) fx.Option {
+	const magicThroughputReportCount = 128
+	return fx.Provide(func(lc fx.Lifecycle, cr irouting.ProvideManyRouter, keyProvider provider.KeyChanFunc, repo repo.Repo, bs blockstore.Blockstore) (provider.System, error) {
+		sys, err := provider.New(repo.Datastore(),
+			provider.Online(cr),
+			provider.ReproviderInterval(reprovideInterval),
+			provider.KeyProvider(keyProvider),
+			provider.ThroughputReport(func(reprovide bool, complete bool, keysProvided uint, duration time.Duration) bool {
+				avgProvideSpeed := duration / time.Duration(keysProvided)
+				count := uint64(keysProvided)
+
+				if !reprovide || !complete {
+					// We don't know how many CIDs we have to provide, try to fetch it from the blockstore.
+					// But don't try for too long as this might be very expensive if you have a huge datastore.
+					ctx, cancel := context.WithTimeout(context.Background(), time.Minute*5)
+					defer cancel()
+
+					// FIXME: I want a running counter of blocks so size of blockstore can be an O(1) lookup.
+					ch, err := bs.AllKeysChan(ctx)
+					if err != nil {
+						logger.Errorf("fetching AllKeysChain in provider ThroughputReport: %v", err)
+						return false
+					}
+					count = 0
+				countLoop:
+					for {
+						select {
+						case _, ok := <-ch:
+							if !ok {
+								break countLoop
+							}
+							count++
+						case <-ctx.Done():
+							// really big blockstore mode
+
+							// how many blocks would be in a 10TiB blockstore with 128KiB blocks.
+							const probableBigBlockstore = (10 * 1024 * 1024 * 1024 * 1024) / (128 * 1024)
+							// How long per block that lasts us.
+							expectedProvideSpeed := reprovideInterval / probableBigBlockstore
+							if avgProvideSpeed > expectedProvideSpeed {
+								logger.Errorf(`
+🔔🔔🔔 YOU MAY BE FALLING BEHIND DHT REPROVIDES! 🔔🔔🔔
+
+⚠️ Your system might be struggling to keep up with DHT reprovides!
+Meaning your content being partially or completely inacessible on the network.
+We observed that you recently provided %d keys at an average rate of %v per key.
+
+🕑 An attempt to estimate your blockstore size timed out after 5 minutes,
+implying your blockstore might be exceedingly large. Assuming a considerable
+size of 10TiB, it would take %v to provide the complete set.
+
+⏰ The total provide time needs to stay under 24 hours to prevent falling behind!
+
+💡 Consider enabling the Accelerated DHT to enhance your system performance. See:
+https://github.com/ipfs/kubo/blob/master/docs/config.md#routingaccelerateddhtclient`,
+									keysProvided, avgProvideSpeed, avgProvideSpeed*probableBigBlockstore)
+								return false
+							}
+						}
+					}
+				}
+
+				// How long per block that lasts us.
+				expectedProvideSpeed := reprovideInterval / time.Duration(count)
```

Comment:
What unit of time are we dealing with here?  Seconds, milliseconds?  I think for clarity, it would be good to embed that into the variable names.

Reply:
The computations are done in nanoseconds, `time.Duration.String` takes care of formating it in something relevent.

Go use a typed time approach, programmers do not have to care about the unit of time, they do operations on the `time.Duration` time (*which internally use nanoseconds*) and then various formating call let extract the precision wanted.